### DISCUSSION TITLE
How to scramble phase/energy with plenoptic?

### DISCUSSION THREAD

**Discussion by billbrod** [plenoptic-org/plenoptic#316] [General]
I was talking with @decrow3 about how to use plenoptic to scramble phase or energy in an image, and thought that the answer would be more useful, so I'm writing it up here:

The way to address this in plenoptic is to build a model that pools the image statistics you want to keep consistent (ignoring those you want to scramble), and use that to generate a metamer. I used this basic approach in [a recent paper](https://elifesciences.org/reviewed-preprints/90554v1) and there's [earlier work](https://dx.doi.org/10.1038/nn.2889) using this from Eero's lab as well.

I'm going to use "pool" here, but you can also think of this as "spatially-weighted averaging" or "blurring" (though I think people more specifically mean "convolve with a Gaussian" when they say "blur", so I'm going to avoid that one).

The basic idea is that you compute a statistic (or feature, if you like) across an image and then average in pre-defined regions. In my paper above, the statistics were either pixel intensity or spectral energy (as computed by the steerable pyramid), and in the earlier paper, the statistics were the [Portilla-Simoncelli texture stats](https://docs.plenoptic.org/docs/branch/main/tutorials/models/Metamer-Portilla-Simoncelli.html).

There are two questions here: which statistics do I pool and how do I pool them? I'm going to leave the "which statistics" question to the side for now. For "how to pool them", you can do in either a spatially-varying or spatially-consistent manner. The two papers I mentioned did so in a spatially-varying manner: the pooling regions were "foveated", laid out in a polar arrangement and growing proportionally to the center of the image. If that's what you want, then https://github.com/plenoptic-org/plenoptic/pull/284 is the branch / PR for you. I'm still working on it, but it is solving this issue for the texture stats specifically, and I hope the machinery to be more generally useful.

The rest of this note will describe how to do spatially-consistent / convolutional pooling. Tools to make this easier will also end up in plenoptic, but hopefully this helps people while I'm working on it.

## Spatially-consistent pooling

If you want to pool your stats in a spatially-consistent manner, that's just convolution. If you look in the beginning of [my tutorial](https://workshops.plenoptic.org/workshops/CSHL-vision-course-2024/branch/main/introduction.html) from the Cold Spring Harbor Computational Vision course last summer, you see an example where I build a simple Gaussian convolutional model:

```python
import torch
import plenoptic as po
# this is a convenience function for creating a simple Gaussian kernel
from plenoptic.simulate.canonical_computations.filters import circular_gaussian2d

# Simple Gaussian convolutional model
class Gaussian(torch.nn.Module):
    def __init__(self, kernel_size=(31, 31), std_dev=3):
        super().__init__()
        self.kernel_size = kernel_size
        self.conv = torch.nn.Conv2d(1, 1, kernel_size=kernel_size, padding=(0, 0), bias=False)
        self.conv.weight.data[0, 0] = circular_gaussian2d(kernel_size, float(std_dev))

    # the forward pass of the model defines how to get from an image to the representation
    def forward(self, x):
        x = po.tools.conv.same_padding(x, self.kernel_size, pad_mode='circular')
        return self.conv(x)

model = Gaussian((31, 31))
rep = model(img)
```

This model takes an image and convolves it with a Gaussian of specified size. `kernel_size` gives the overall size of the kernel, while `std_dev` gives the standard deviation of the kernel. `std_dev` is thus the parameter of interest, with larger values making the pooling region larger (and `kernel_size` should be large enough that the filter goes roughly to zero). We also use circular padding to make sure the output and the input are the same size. How you handle padding will influence the presence of border artifacts, so be aware.

After running the code above, `model` can be used to generate metamers as shown [in the docs](https://docs.plenoptic.org/docs/branch/main/tutorials/intro/06_Metamer.html) (or later in that CSHL tutorial).

In terms of the framework I laid out above, `Gaussian` pools pixel values in a spatially consistent manner across the image. If you wanted it to pool some other statistic, you would modify the `forward` function, computing something else that you would pass to `self.conv`. For example, if you wanted to do this on energy from low frequencies:

```python
import torch
import plenoptic as po
# this is a convenience function for creating a simple Gaussian kernel
from plenoptic.simulate.canonical_computations.filters import circular_gaussian2d

class GaussianLowFreqs(torch.nn.Module):
    def __init__(self, kernel_size=(31, 31), img_size=(256, 256), std_dev=3):
        super().__init__()
        self.kernel_size = kernel_size
        self.pyr = po.simul.SteerablePyramidFreq(img_size, 1, is_complex=True, tight_frame=True)
        self.conv = torch.nn.Conv2d(1, 1, kernel_size=kernel_size, padding=(0, 0), bias=False,)
        self.conv.weight.data[0, 0] = circular_gaussian2d(kernel_size, float(std_dev))

    # the forward pass of the model defines how to get from an image to the representation
    def forward(self, x):
        pyr_coeffs = self.pyr(x)
        # ignore the residuals
        pyr_coeffs = {k: po.tools.conv.same_padding(v, self.kernel_size, pad_mode='circular')
                      for k, v in pyr_coeffs.items() if not isinstance(k, str)}
        energy = torch.cat([v.pow(2).abs() for v in pyr_coeffs.values()], 0)
        return self.conv(energy)

model = GaussianLowFreqs((31, 31))
rep = model(img)
```

Note that the above puts the pyramid's orientation on the batch dimension; it will probably not behave as expected if you have a multi-batch or multi-channel image.

## Multi-scale pooling

If you have a multi-scale representation, such as multiple scales from the pyramid, you'll need to have the pooling kernel defined for each scale (convolving the same kernel with each scale is equivalent to growing the pooling region for each coarser scale). Eventually, plenoptic will have a component to handle this, which will operate similar to the `WeightedAveragePyramid` in https://github.com/plenoptic-org/plenoptic/pull/284.

It would look something like:

```python
import torch
from plenoptic.simulate.canonical_computations.filters import circular_gaussian2d
class ConvolutionalPyramid(torch.nn.Module):
    def __init__(self, kernel: torch.Tensor, n_scales: int = 4):
        super().__init__()
        self.n_scales = n_scales
        kernels = []
        for i in range(n_scales):
            if i != 0:
                kernel = po.tools.blur_downsample(kernel, 1, scale_filter=True)
            conv = torch.nn.Conv2d(1, 1, kernel_size=kernel.shape[-2:], padding=(0, 0), bias=False)
            conv.weight.data[0, 0] = kernel.clip(min=0)
            kernels.append(conv)
            self.kernels = torch.nn.ModuleList(kernels)

    def forward(self, x: list[torch.Tensor]) -> list[torch.Tensor]:
        """Pool a list of multi-scale images

        Parameters
        ----------
        x :
            A list of `n_scales` 4d tensors, each of which has been downsampled by 2 (so
            that their heights halve each time).

        """
        x = [po.tools.conv.same_padding(x_scale, kernel.kernel_size, pad_mode='circular')
             for x_scale, kernel in zip(x, self.kernels)]
        return [kernel(x_scale) for x_scale, kernel in zip(x, self.kernels)]


class EnergyPyramidGaussian(torch.nn.Module):
    def __init__(self, kernel_size=(31, 31), img_size=(256, 256), n_scales=4, n_orientations=4,
                 std_dev=3):
        super().__init__()
        self.kernel_size = kernel_size
        self.n_scales = n_scales
        self.pyr = po.simul.SteerablePyramidFreq(img_size, n_scales, order=n_orientations-1,
                                                 is_complex=True, tight_frame=True)
        self.conv = ConvolutionalPyramid(circular_gaussian2d(kernel_size, float(std_dev)),
                                         n_scales)

    def forward(self, x):
        pyr_coeffs = self.pyr(x)
        # convert pyr coeffs into a list of Tensors, ignoring residuals
        coeffs_list = [torch.cat([pyr_coeffs[(i, j)] for j in range(self.pyr.order+1)], 0)
                       for i in range(self.n_scales)]
        energy = [v.pow(2).abs() for v in coeffs_list]
        pooled_energy = self.conv(energy)
        # plenoptic requires models return a single 3d or 4d tensor, so flatten and concatenate
        # across scales.
        return torch.cat([v.flatten(-2, -1) for v in pooled_energy], -1)

model = EnergyPyramidGaussian((31, 31))
rep = model(img)
```

Note that the above puts the pyramid's orientation on the batch dimension; it will probably not behave as expected if you have a multi-batch or multi-channel image.

If `EnergyPyramidGaussian` is run on an image of shape `(1, 1, 256, 256)`, it will return a list of `n_scales` tensors, each of shape `(n_orientations, 1, 256/2**i, 256/2**i)`, where `n_scales`, `n_orientations` are set at model initialization, and `i` runs from 0 to `n_scales-1`.

Note there are probably other ways to do the final cat/flatten operation to make the representation easier to investigate.

## Aliasing

You can play with the spacing of these pooling regions by increasing the stride of the `torch.nn.Conv2d` module (see [its docs](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html)).

However, you should be concerned about aliasing: if your stride is too large, you will introduce [aliasing](https://en.wikipedia.org/wiki/Aliasing) into your representation. You can check against this in several ways:

- You should be able to use linear interpolation to reconstruct your kernel at the locations you are **not** sampling. (See [this notebook](https://github.com/LabForComputationalVision/pooling-windows/blob/main/Sampling_and_Aliasing.ipynb) for an example).
- Pass an image through your pooling regions, then take the pseudo-inverse to reconstruct your initial image. Shift the regions by half the sampling distance, then do it again. Your reconstruction shouldn't change much.

For any reasonable kernel, sampling at every pixel (i.e., `stride=1`) should work, and there should be some distance large enough where aliasing arises and the above checks fail. So if you run the above checks while gradually increasing the sampling distance, you should see everything is fine until suddenly it's not.

This distance will depend on your kernel, so you should first pick the functional form of your pooling region, and then run the above checks to find the sampling distance / stride.

For Gaussian windows, in my [foveated metamer paper](https://elifesciences.org/reviewed-preprints/90554v1), I convinced myself that a sampling of half the standard deviation had no aliasing.
